src/Diagnostics/Debug/StateCheck.jl

```diff
@@ -0,0 +1,604 @@
+Q1=MPIStateArray{Float32,F1}(MPI.COMM_WORLD,ClimateMachine.array_type(),4,9,8);
+Q2=MPIStateArray{Float64,F2}(MPI.COMM_WORLD,ClimateMachine.array_type(),4,6,8);
+cb=ClimateMachine.StateCheck.sccreate([(Q1,"My gradients"),(Q2,"My fields")],1; prec=$prec_def);
+cb()
+```
+"""
+function sccreate(
+    io::IO,
+    fields::Array{<:Tuple{<:MPIStateArray, String}},
+    nt_freq::Int = nt_freq_def;
+    prec = prec_def,
+)
+    if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+        println(io, "# SC Start: creating state check callback")
+    end
+
+    ####
+    # Print fields that the call back create by this call will query
+    ####
+    nr = 0
+    for f in fields
+        print_head = true
+        Q = f[1]
+        lab = f[2]
+        V = typeof(Q).parameters[2]
+        slist = typeof(Q).parameters[2].names
+        l_s = length(slist)
+        nss = 0
+        if l_s == 0
+            println(
+                io,
+                "# SC  MPIStateArray labeled \"$lab\" has no named symbols.",
+            )
+        else
+            ns = 0
+            for s in slist
+                ns = ns + 1
+                if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+                    if print_head
+                        println(
+                            io,
+                            "# SC Creating state check callback labeled \"$lab\" for symbols",
+                        )
+                    end
+                    print_head = false
+                    println(io, "# SC ", s)
+                end
+                for n in
+                    flattenednames(fieldtype(V, ns), prefix = fieldname(V, ns))
+                    nss = nss + 1
+                end
+            end
+        end
+        nr = nr + nss
+    end
+
+    ###
+    # Initialize total calls counter for the call back
+    ###
+    n_cb_calls = 0
+
+    ###
+    # Create holder for most recent stats
+    ###
+    cur_stats_dict = Dict()
+    cur_stats_flat = Array{Any}(undef, nr)
+
+    # Save io pointer
+    iosave = io
+
+    ######
+    # Create the callback
+    ######
+    cb = EveryXSimulationSteps(nt_freq) do (s = false)
+        # Track which timestep this is
+        n_cb_calls = n_cb_calls + 1
+        n_step = (n_cb_calls - 1) * nt_freq + 1
+        ns_str = @sprintf("%7.7d", n_step - 1)
+        io = iosave
+        # Obscure trick to do with running in cells in notebook that close Base.stdout
+        # between different stages. This affects parsing Literate/Documenter examples.
+        if !isopen(io)
+            io = Base.stdout
+        end
+
+        ## Print header
+        nprec = min(max(1, prec), 20)
+        if MPI.Comm_rank(MPI.COMM_WORLD) == 0
+            println(
+                io,
+                "# SC +++++++++++ClimateMachine StateCheck call-back start+++++++++++++++++",
+            )
+            println(
+                io,
+                "# SC  Step  |   Label    |  Field   |                            Stats                       ",
```

Comment:
Does the output have to be formatted text? Writing JLD2s or NC files would make both writing the output and reading it back much easier.

Reply:
@christophernhill : Can you please fill in the "Description" section for the PR with a summary of the feature being added.

Reply:
> Does the output have to be formatted text? Writing JLD2s or NC files would make both writing the output and reading it back much easier.

The plain text format is deliberate. Generally for lightweight stuff like this a human readable little fragment can work quite well. For example https://github.com/CliMA/ClimateMachine.jl/blob/dd706a4ac370db3499afc5b3a8459d3d11f70992/test/Numerics/LinearSolvers/poisson.jl#L234 does this and we have been trialing this in https://github.com/christophernhill/temp-clima-statetools/blob/master/test/Ocean/HydrostaticBoussinesq/test_ocean_gyre_refvals.jl and finding it works well. 

Importantly the direct human manipulation format is designed to allow expert knowledge to be edited into the precision match numbers. For different fields and different statistics and different tests the statistical match needs to be set appropriately.